storage/filesystem/object.go

```diff
@@ -294,25 +298,29 @@ func newPackfileIter(f billy.File, t plumbing.ObjectType, seen map[plumbing.Hash
 }
 
 func (iter *packfileIter) Next() (plumbing.EncodedObject, error) {
-	if iter.position >= iter.total {
-		return nil, io.EOF
-	}
+	for {
+		if iter.position >= iter.total {
+			return nil, io.EOF
+		}
 
-	obj, err := iter.d.DecodeObject()
-	if err != nil {
-		return nil, err
-	}
+		obj, err := iter.d.DecodeObject()
+		if err != nil {
+			return nil, err
+		}
 
-	iter.position++
-	if iter.seen[obj.Hash()] {
-		return iter.Next()
-	}
+		iter.position++
+		if obj != nil {
+			if iter.seen[obj.Hash()] {
+				return iter.Next()
+			}
 
-	if iter.t != plumbing.AnyObject && iter.t != obj.Type() {
-		return iter.Next()
-	}
+			if iter.t != plumbing.AnyObject && iter.t != obj.Type() {
+				return iter.Next()
```

Comment:
a recursive call inside a for loop? on a condition that can not happen because the code just added to the file above? WTF.